proposals/p1928.md

```diff
@@ -0,0 +1,163 @@
+# Arrays
+
+<!--
+Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+Exceptions. See /LICENSE for license information.
+SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+-->
+
+[Pull request](https://github.com/carbon-language/carbon-lang/pull/1928)
+
+<!-- toc -->
+
+## Table of contents
+
+-   [Problem](#problem)
+-   [Background](#background)
+-   [Proposal](#proposal)
+-   [Details](#details)
+-   [Rationale](#rationale)
+-   [Alternatives considered](#alternatives-considered)
+
+<!-- tocstop -->
+
+## Problem
+
+Arrays are base component of almost of programming languages. In the current
+stage of Carbon development, arrays are missed.
+
+## Background
+
+An array is a set of scalar data placed in contiguous memory locations, all of
+the same type and type parameters. The scalar data that make up an array are the
+array elements which can be referenced by unique indexes.
+
+A nested array is an array of arrays which may be non-continuous in memory.
+
+A multidimensional array is an array with more than two dimensions which is
+continuous in memory.
+
+## Proposal
+
+This proposal introduces array internal representation, its declaration, and
+indexing.
+
+## Details
+
+### Internal representation
+
+For simplifying usage and raising safety of array usage, internal representation
+of arrays contains additional metadata:
+```cpp
+  struct {
+    int32_t dimensions;
```

Comment:
Nit. I could definitely see some applications having 2gb character strings and this size being insufficient. A `uintptr_t` would work by definition.

But I'm also hesitant to have this in optimized builds. I don't know what the cost is, but it's certainly not free. Should this be only in debug and hardening modes?

Reply:
`dimensions` variable contains only number of dimensions. For example, `1` for 1D array and `2` for 2D array. It is the size of the following array. Anyway, you raised the problem and it is actual for the following `dimension_size` variable.

BTW, what is the problem with optimized builds? Does it different from `void func(void * data, int32_t data_len);`? Anyway, you should pass the length of your array (strings usually contains "\0", so we know where is the end).

Reply:
I think the principle I'm relying on here is that, for optimized builds, we should make programmer errors undefined behavior so that the compiler is free to optimize assuming they do not exist. Certainly I see the value in catching these bugs in debug and hardened modes, but if I cannot remove debug information in optimized builds, we have left room for a lower level, more efficient language, failing at a core principle of the language.

For arrays whose length is known at compile-time, it is a programmer error to access the array out of bounds and so, for optimized builds, should be undefined behavior. Thus, for such arrays, there is no need to store the dimensions at all.